test/framework/src/main/java/org/elasticsearch/index/engine/EngineTestCase.java

```diff
@@ -673,30 +710,119 @@ public static void assertOpsOnReplica(
-        final Set<Long> seqNos = new HashSet<>();
+        final List<Translog.Operation> operations = new ArrayList<>();
         try (Engine.Searcher searcher = engine.acquireSearcher("test", Engine.SearcherScope.INTERNAL)) {
-            IndexSearcher indexSearcher = new IndexSearcher(Lucene.wrapAllDocsLive(searcher.getDirectoryReader()));
-            List<LeafReaderContext> leaves = indexSearcher.getIndexReader().leaves();
-            NumericDocValues[] seqNoDocValues = new NumericDocValues[leaves.size()];
-            for (int i = 0; i < leaves.size(); i++) {
-                seqNoDocValues[i] = leaves.get(i).reader().getNumericDocValues(SeqNoFieldMapper.NAME);
-            }
-            TopDocs allDocs = indexSearcher.search(new MatchAllDocsQuery(), Integer.MAX_VALUE);
+            final IndexSearcher indexSearcher = new IndexSearcher(Lucene.wrapAllDocsLive(searcher.getDirectoryReader()));
+            final Sort sortedBySeqNoThenByTerm = new Sort(
+                new SortedNumericSortField(SeqNoFieldMapper.NAME, SortField.Type.LONG),
+                new SortedNumericSortField(SeqNoFieldMapper.PRIMARY_TERM_NAME, SortField.Type.LONG, true)
+            );
+            final TopDocs allDocs = indexSearcher.search(new MatchAllDocsQuery(), Integer.MAX_VALUE, sortedBySeqNoThenByTerm);
+            long lastSeenSeqNo = SequenceNumbers.NO_OPS_PERFORMED;
             for (ScoreDoc scoreDoc : allDocs.scoreDocs) {
-                int leafIndex = ReaderUtil.subIndex(scoreDoc.doc, leaves);
-                int segmentDocId = scoreDoc.doc - leaves.get(leafIndex).docBase;
-                if (seqNoDocValues[leafIndex] != null && seqNoDocValues[leafIndex].advanceExact(segmentDocId)) {
-                    seqNos.add(seqNoDocValues[leafIndex].longValue());
-                } else {
-                    throw new AssertionError("Segment without seqno DocValues");
+                final Translog.Operation op = readOperationInLucene(indexSearcher, mapper, scoreDoc.doc);
+                if (op.seqNo() != lastSeenSeqNo) {
+                    operations.add(op);
+                    lastSeenSeqNo = op.seqNo();
                 }
             }
         }
-        return seqNos;
+        return operations;
+    }
+
+    private static Translog.Operation readOperationInLucene(IndexSearcher searcher, MapperService mapper, int docID) throws IOException {
+        final List<LeafReaderContext> leaves = searcher.getIndexReader().leaves();
+        final int leafIndex = ReaderUtil.subIndex(docID, leaves);
+        final int segmentDocID = docID - leaves.get(leafIndex).docBase;
+        final long seqNo = readNumericDV(leaves.get(leafIndex), SeqNoFieldMapper.NAME, segmentDocID);
+        final long primaryTerm = readNumericDV(leaves.get(leafIndex), SeqNoFieldMapper.PRIMARY_TERM_NAME, segmentDocID);
+        final FieldsVisitor fields = new FieldsVisitor(true);
+        searcher.doc(docID, fields);
+        fields.postProcess(mapper);
```

Comment:
do we really need this? if not, then we can avoid chaining in the mapper..

Reply:
Yeah, I can get docType explicitly and remove this call.

Reply:
@bleskes We have to call `postProcess` to extract doc type and doc id (via Uid); otherwise we have to pass docType into FieldsVisitor manually.